documents/extension/funding/service.go

```diff
@@ -0,0 +1,265 @@
+	key, err := documents.AttrKeyFromLabel(fundingLabel)
+	if err != nil {
+		return idx, err
+	}
+
+	attr, err := model.GetAttribute(key)
+	if err != nil {
+		return idx, err
+	}
+
+	idx = attr.Value.Int256
+
+	z, err := documents.NewInt256("0")
+	if err != nil {
+		return idx, err
+	}
+
+	// idx < 0
+	if idx.Cmp(z) == -1 {
+		return idx, ErrFundingIndex
+	}
+
+	return idx, nil
+
+}
+
+func defineFundingAttrIdx(model documents.Model) (attr documents.Attribute, err error) {
+	key, err := documents.AttrKeyFromLabel(fundingLabel)
+	if err != nil {
+		return attr, err
+	}
+
+	if !model.AttributeExists(key) {
+		return documents.NewAttribute(fundingLabel, documents.AttrInt256, "0")
+	}
+
+	idx, err := getFundingsLatestIdx(model)
+	if err != nil {
+		return attr, err
+	}
+
+	// increment idx
+	newIdx := idx.Inc()
+
+	if err != nil {
+		return attr, err
+	}
+
+	return documents.NewAttribute(fundingLabel, documents.AttrInt256, newIdx.String())
+
+}
+
+func createAttributesList(current documents.Model, data Data) ([]documents.Attribute, error) {
+	var attributes []documents.Attribute
+
+	idx, err := defineFundingAttrIdx(current)
+	if err != nil {
+		return nil, err
+	}
+
+	// add updated idx
+	attributes = append(attributes, idx)
+
+	types := reflect.TypeOf(data)
+	values := reflect.ValueOf(data)
+	for i := 0; i < types.NumField(); i++ {
+		jsonKey := types.Field(i).Tag.Get("json")
+		label := labelFromJSONTag(idx.Value.Int256.String(), jsonKey)
+		
+		value := values.Field(i).Interface().(string)
+		attrType := types.Field(i).Tag.Get("attr")
+
+		attr, err := documents.NewAttribute(label, documents.AttributeType(attrType), value)
+		if err != nil {
+			return nil, err
+		}
+
+		attributes = append(attributes, attr)
+
+	}
+
+	return attributes, nil
+}
+
+func (s service) DeriveFromPayload(ctx context.Context, req *clientfundingpb.FundingCreatePayload, identifier []byte) (documents.Model, error) {
+	fd := Data{}
+	fd.initFundingFromData(req.Data)
+
+	model, err := s.GetCurrentVersion(ctx, identifier)
+	if err != nil {
+		apiLog.Error(err)
+		return nil, centerrors.Wrap(err, "document not found")
+	}
+
+	attributes, err := createAttributesList(model, fd)
+	if err != nil {
+		return nil, err
+	}
+
+	err = model.AddAttributes(attributes...)
+	if err != nil {
+		return nil, err
+	}
+
+	validator := CreateValidator()
+	err = validator.Validate(nil, model)
+	if err != nil {
+		return nil, errors.NewTypedError(documents.ErrDocumentInvalid, err)
+	}
+
+	return model, nil
+}
+
+func (s service) findFunding(model documents.Model, fundingID string) (idx string, err error) {
+	lastIdx, err := getFundingsLatestIdx(model)
+	if err != nil {
+		return idx, err
+	}
+
+	i, err := documents.NewInt256("0")
+	if err != nil {
+		return idx, err
+	}
+
+	for ; i.Cmp(lastIdx) != 1; i.Inc() {
+		label := generateLabel(i.String(), fundingIDLabel)
+		k, err := documents.AttrKeyFromLabel(label)
+		if err != nil {
+			return idx, err
+		}
+
+		attr, err := model.GetAttribute(k)
```

Comment:
As I understand it, here there is an iteration over all the fundings stored to find the correct one. Wouldnt this GetAttribute return error when is not found? and therefore never keep looking for more until the fundingIdx is matched?

Reply:
Yes, it would return an error.

Do you mean it is not necessary to store the length of the array?
I think it could be useful to have it.

Reply:
Why do we iterate then? if for more than 1 entry we will never look forward? That is my only question.

Reply:
A document could have multiple funding agreements.

The `funding_id` is a value and not a key. We might have 10 fundings but we don't know which index has the funding with the `funding_id` we are looking for.

If we would store it like `centrifuge_funding[{funding_id}]` it would not be possible to return a list of all funding agreements.